File: app/src/main/java/com/vo1d/schedulemanager/v2/data/subjects/SubjectDao.java
package com.vo1d.schedulemanager.v2.data.subjects;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.Transaction;

import com.vo1d.schedulemanager.v2.data.IBaseDao;

import java.util.List;

@Dao
public abstract class SubjectDao implements IBaseDao<Subject> {

    @Transaction
    @Query("DELETE FROM subject_table")
    abstract void deleteAll();

    @Transaction
    @Query("SELECT * FROM subject_table ORDER BY title ASC")
    abstract LiveData<List<Subject>> getAll();

    @Transaction
    @Query("SELECT * FROM subject_table WHERE title LIKE '%' || :filter || '%' ORDER BY title ASC")
    abstract List<Subject> getFilteredSubjects(String filter);

    @Query("SELECT * FROM subject_table WHERE id=:id")
    abstract Subject findSubjectById(int id);
}

File: core/data/src/main/kotlin/ru/vo1dmain/timetables/data/entities/subject/SubjectDao.kt
package ru.vo1dmain.timetables.data.entities.subject

import androidx.room.Dao
import androidx.room.Query
import androidx.room.Transaction
import kotlinx.coroutines.flow.Flow
import ru.vo1dmain.timetables.data.entities.BaseDao

@Dao
internal interface SubjectDao : BaseDao<SubjectEntity> {
    @Transaction
    @Query("SELECT * FROM subjects")
    fun getAll(): Flow<List<SubjectWithTeachers>>
    
    @Transaction
    @Query(
        """SELECT * FROM subjects
        WHERE title LIKE '%' || :filter || '%'
        ORDER BY title ASC"""
    )
    suspend fun filter(filter: String): List<SubjectWithTeachers>
    
    @Transaction
    @Query("SELECT * FROM subjects WHERE id = :id")
    suspend fun findById(id: Int): SubjectWithTeachers?
    
    @Query("DELETE FROM subjects WHERE id IN (:ids)")
    suspend fun deleteByIds(ids: Collection<Int>)
}